### ISSUE TITLE
Make DCM4CHE and all dependencies available on maven central

### ISSUE THREAD

**Issue by AlexanderSchuetz97** [dcm4che/dcm4che#1367]
**Is your feature request related to a problem? Please describe.**
My Problem is, that everytime dcm4chee releases an update I have to download the binary release,
find all required jar files from the lib folder and manually upload each one to a self hosted maven repository/mirror.
This wastes a lot of time.
The main artifacts I have to upload are usually
- dchm4chee-XXXXXXX-VERSION.jar
- weasis-core-img-VERSION.jar
- libopencv .dll/.so (Ideally this would be also bundled in a jar file)

**Describe the solution you'd like**
Please upload all artifacts to maven central. 
This task can be almost entirely automated by the maven build system you are already using.

**Describe alternatives you've considered**
I could fully automate uploading the artifacts from the binary release on my end, this is however prone to errors since
the dependencies/names etc of the artifacts can change with every release on your end.

**Additional context**
I am using dcm4che as a java library.
As I am sure you are aware:
Most modern java development setups use libraries from a maven repository. 
Build tools such as gradle/maven download those dependencies from there.
Since most modern software is also build on a CI, potentically in containers, 
not having the artifacts in some repository is not really feasable anymore.

**Comment by therealryan** [comment]
While the dcm4che artifacts are not available on maven central, they _are_ available from [scijava.org](https://maven.scijava.org/content/repositories/public/org/dcm4che/). Add a `repositories` block to your pom file like so:
```xml
<project>
	<repositories>
		<repository>
			<!-- source for dcm4che artifacts -->
			<id>scijava</id>
			<url>https://maven.scijava.org/content/repositories/public/</url>
		</repository>
	</repositories>
</project>
```
then you can declare your dependencies as normal:
```xml
<project>
	<dependencies>
		<dependency>
			<groupId>org.dcm4che</groupId>
			<artifactId>dcm4che-core</artifactId>
			<version>5.31.2</version>
		</dependency>
	</dependencies>
</project>
```

It looks like the weasis lib is [also available](https://maven.scijava.org/content/repositories/public/com/github/nroduit/) from that source, but I'm not sure about the dll file you're after.

**Comment by AlexanderSchuetz97** [comment]
@therealryan hello
if you want to use dcm4che-imageio/opencv variant you will need to use "weasis-core-img"
The only known way to me is to get dcm4che-imageio/opencv working is to build weasis-core-img from source.
The last versions of weasis lib available on any repository is 4.5. We currently need version 4.8.1. The weasis lib needs several native components. (.dll/.so files) that are loaded by JNI and have to present in the native library path of the application. 

As you can imagine the pain is high for me. I had to do this yet again today for dcm4che 5.31.2...

1. Find the source code of weasis lib for the version 4.8.1. (Which is non trivial as the guy who maintains this library doesnt have any releases on his github page so I have to find the last commit where pom.xml contains 4.8.1 and pray that that is the one that will actually work)
2. Setup the correct JDK and the correct java version and correct maven version (thankfully docker exists)
3. Compile the java part (.jar file)
4. Serach for a random dicom viewers on the internet to find the native components of weasis lib for windows and linux. (Compiling these would take several hours as I essentially would have to setup a VM just compile this and reverse engineer the compilation instructions since there is 0 documentation)

I presume someone of you guys has a version of all of this somewhere that just works, this recall cant be that hard...
I don't care where you upload it. (Maven Central would still be the easiest way to do it, just saying) Just put it all in 1 place where an automated tool can find it. I was already aware of your
internal maven repo but for a while it seemed to be abandoned. You seemingly have revived it again.

**Comment by gunterze** [comment]
weasis maven repository: https://github.com/nroduit/mvn-repo/tree/master
```xml
    <repository>
      <snapshots>
        <enabled>false</enabled>
      </snapshots>
      <id>mvn-repo-master</id>
      <url>https://raw.github.com/nroduit/mvn-repo/master/</url>
    </repository>
 ```

**Comment by AlexanderSchuetz97** [comment]
@gunterze Thank you for this information. 
Perhaps you should list this together with dcm4chee's mvn repo as the second 
point in the main README of this repository.
Having this info more accessible would have saved me a lot of time and headaches.

I get the impression that you and the rest of the dcm4chee contributors are for some reason that I do not know or cannot guess not interested in putting this on maven central. If this is indeed so then I would be happy if you could tell me your reasoning and then you can If you want, close this issue.

**Comment by hczedik** [comment]
I would also be very much in favor of having the complete dcm4che library on Maven Central, including all the dependencies.
I believe that not having it there, is an obstacle for a more popular adoption of the library and I shared that concern with @gunterze already before.  In general, I believe having the library available on Maven central would make Java+dcm4che an even more popular choice for the larger DICOM community. I do believe that dcm4che is one of best choices for working with DICOM data and services. The main pain points, especially for new users, in my opinion, currently are the difficult availability and also a bit of lack of good documentation. That is why, I believe, a lot of new users rather opt for other library choices (like Pydicom), even if they would be already familiar with the Java programming language.

**Comment by nroduit** [comment]
I'm not sure that putting all dependencies on Maven Central would solve the problem you mention as all libraries are already available in these 2 repositories:
``` xml
    <repository>
      <id>mvn-repo-master</id>
      <url>https://raw.github.com/nroduit/mvn-repo/master/</url>
      <snapshots>
        <enabled>false</enabled>
      </snapshots>
    </repository>
    <repository>
      <id>www.dcm4che.org</id>
      <name>dcm4che Repository</name>
      <url>https://www.dcm4che.org/maven2</url>
    </repository>
```

> The last versions of weasis lib available on any repository is 4.5. We currently need version 4.8.1. 

=> You don't need to compile this library which is [available in this repository](https://github.com/nroduit/mvn-repo/tree/master/org/weasis/core/weasis-core-img).

We're also faced with the problem of managing dependencies by linking libraries that themselves have references to other libraries (as in Weasis, which integrates modules from dcm4che with sometimes a more recent version of weasis-core-img). To simplify management, we start to use a [BOM type](https://reflectoring.io/maven-bom/) to manage transitive dependencies and version consistency more simply. For example, all you need to do is define the version of [weasis-core-img ](https://github.com/nroduit/weasis-core-img/blob/master/pom.xml)to automatically get the right version of the native part. We will soon be proposing a PR to integrate a BOM import of weasis-core-img and native libraries in dcm4che.

For your information, the libraries dcm4che, weasis-core-img and its native part will always have different versions, as they have their own life cycle and are released by different people.

**Comment by jssuttles** [comment]
On the topic of dependencies: https://github.com/dcm4che/dcm4che/issues/1416

**Comment by AlexanderSchuetz97** [comment]
@nroduit 
Hello

Your link does only link to the jar files which is already a big help. 1 step less to worry about.
Where does one find the compiled shared objects for linux amd64 and windows amd64/i386?

They are not inside of the jars in this link, ive unzipped them. I do not want to once again have to google for random dicom viewers that use this dll to find it. Can you please link it for me?

Also a BOM does not improve my problem. The transitive dependencies are already setup correctly and know the version of weasis-core-img that I need. With the knowlege that a maven repo for weasis exists, I can already skip the compilation of the jar part. I still need the so/dll files. 

The main reason why maven central would make it easy is because you wouldnt need this insider knowlege that there is a random maven repo inside your git repo where you can download this stuff. A person would just put dcm4chee imagio in whatever version into his pom (Thats what most people want dcm file to BufferedImage) and everything would just work. 

Except for the native part which you "could" fix by including the shared objects as resources inside your jar file, many libraries like the pytorch bindings for java for example do it that way. The strategy is generally to unpack the lib into the temp dir and then call System.load on it. You currently call loadLibrary in a static block and one just has to pray that the correct shared object is in the native lib path. I have workarounds for this problem in place but this is another area that could be made smoother.

TLDR
Give me a link to your native binaries please.

Sincerely
Alexander Schütz

**Comment by nroduit** [comment]
Thank you for describing the problem you're experiencing in detail because I have the impression that we're not talking about the same problems.

As a reminder, the wrapper part or the [native part](https://github.com/nroduit/weasis-core-img/blob/master/pom.xml#L60-L110) of weasis-core-img is directly [available as binaries](https://github.com/nroduit/mvn-repo/tree/master/org/weasis/thirdparty/org/opencv) in the Maven repo. The advantage of using the BOM in dcm4che is that it references more than just the jar version. The version of the native part is set automatically to guarantee consistency between the two (e.g. [jar 4.9.0.1](https://github.com/dcm4che/dcm4che/blob/master/pom.xml#L84) which automatically references the [native 4.9.0-dcm](https://github.com/dcm4che/dcm4che/blob/e18a00356ace75dac885d215c437fb6635570d92/dcm4che-jboss-modules/pom.xml#L166-L207) version). 

I'm not sure I understand the last paragraph. There's no universal way to deploy a native library in Java, because it depends on the container (Docker, JBoss, Tomcat...) and the system. In our various software applications, this is why we have different integration and deployment management implementations, depending on the context.

**Comment by AlexanderSchuetz97** [comment]
@nroduit 
Sorry for not answering your question for a long time. I forgot about this.

The "universal" way to deploy a native library in Java that always works is by writing a unpacker.
The unpacker's job is essentially to extract a resource dll/so from the jar/classpath/classloader and put it somewhere on the file system. Usually the tmpdir. Some libraries extract it to USER_HOME instead. It really does not matter much. The deployment via System.loadLibrary("...") and relying on dev-ops to put the dll/so into the native path directory is in my humble opinion the worst way to do it as it requires the user of the library to perform active tasks to correctly load the library. (You have to 1. find the so/dll, 2. put it in a folder 3. set a -D option to start the jvm...)

An example where I have implemented this would be this Class I have written for a library of mine that has a native component:
https://github.com/AlexanderSchuetz97/JavaNativeUtils/blob/master/JavaNativeUtilsJni/src/main/java/eu/aschuetz/nativeutils/impl/NativeLibraryLoaderHelper.java

I would say the majority of libraries with a native component that were implemented in the last 10 years do it this way or a similar way. (Pytorch (djl), newer versions of RXTX, jetty for http3)

**Comment by svaens** [comment]
Hi all, 
Sorry in advance  - this is I think strictly speaking the wrong forum for this question. However, as it is related, and I don't know where I would more likely get an answer from an informed person, I charge on.

I was just now in the process of migrating a couple small java DICOM applications (SCU,.. small SCP)  from an old proprietary library, to ... what I hoped would be a more modern, better maintained, open source library. And so i'm looking at dcm4che.

I started with my exploratory fiddling on version 5.26.1
Then, decided to try (i'm currently fixed on Java 8 unfortunately) the latest 5.33.1.
To my surprise and dismay, I find suddenly dependencies won't resolve - it's because its trying to pull in Weasis. 

To my understanding, **Weasis is a DICOM viewer.** 

I've no interest in this, as we have our own DICOM viewer. 

So, is this a mistake? Did someone leave a dependency in accidentally? 

Unlikely, eh? because... now I notice that it has been the case since version 5.31.3
(Feb 06, 2024)

Googling this new dependency doesn't seem to find much at all except for this issue.

If someone can give me an idea of where to go from here .. it would be appreciated. 

Best,

Sean

**Comment by svaens** [comment]
@AlexanderSchuetz97  Hi! And thanks for your response! 
And, yes! I would be very interested in hearing about your solutions. I'll send you an email asap.
